### ISSUE TITLE
Improve Openess Score so information architecture focus on empty space and liberation from binary blobs

### ISSUE THREAD

**Issue by pietrushnic** [Dasharo/dasharo-issues#924]
- [ ] Change the architecture of information in the output so free space is promoted; we can use the discussion from this thread to write a justification for that thinking. In general, my opinion is similar to Ron's: about firmware getting out of the way as fast as possible, which means a smaller code base serving the purpose of users is better than a more extensive code base that is fully featured. This could lead to the conclusion of the direction of innovation in the firmware space, which should happen. Here is another exciting perspective on the same problem: https://world.hey.com/dhh/finished-software-8ee43637, which sticks with KISS and UNIX design ideas.
- [ ] Code that can generate a comparison table of various firmware already scored on the Dasharo page; I could present that on the next DUG because it could be interesting for the community.

_Originally posted by @pietrushnic in https://github.com/Dasharo/dasharo-issues/issues/917#issuecomment-2199869749_

**Comment by miczyg1** [comment]
> Change the architecture of information in the output so free space is promoted; we can use the discussion from this thread to write a justification for that thinking.

Maybe we could simply introduce "liberation factor" calculated with the formula:

(open source bytes + empty bytes) x 100% / closed source bytes. That way we promote empty space in favor of open source code. We can add an exception: if open source bytes  == 0 then liberation factor = 0%. I don't think overcomplicating it will benefit us at all.

> Code that can generate a comparison table of various firmware already scored on the Dasharo page

Looks like out of scope of what the utility does and is supposed to do. Sounds more like another utility to gather the data from docs.dasharo.com. Also does not seem to be related directly to `focus on empty space and liberation from binary blobs` (a separate issue for that would be better).

What exactly we would like to compare? Just the raw percentage of open/clsoed code ratio/liberation factor? Openness Score report contains much more data which does not necessarily needs to be compared, nor will give meaningful data. For now I don;t see what information could such comparison table provide, except openness/liberation percentage or raw closed source size change over the years of CPU evolution.

Another matter to solve is where to gather the data input for these comparisons. It should be something that is easily parsable and can be quickly imported to an array or something. Simplest would be CSV file...

Feeding it with the same platforms over and over does not increase the variety of results (e.g. feeding the utility with multiple releases of the same platform would just give neglectable differences).

@pietrushnic

**Comment by miczyg1** [comment]
> How do you visualize that? We should experiment because I need vision our of the box.

I'm an engineer, I can only think about providing raw data. The representation layer is not my forte. Whatever we choose, I would probably have problems to give opinion if it is good or bad... Honestly, I don't know how to show that, because it is mostly personal view. One will like it, other not, and the cycle of changes never ends.

**Comment by pietrushnic** [comment]
Let's try what I wrote here:

https://github.com/Dasharo/dasharo-issues/issues/276#issuecomment-2548193678

> The pie chart may be the same as before, but it is shown differently (side by side), so when you look at it, you can visually compare what changed in what area. A pie chart with some dotted lines shows how things looked before, and solid lines show the current state.

Engineers have to present results. Drawing charts with good legends is essential for mass consumption.

**Comment by miczyg1** [comment]
> Engineers have to present results.

Yes, they do that very well with raw numbers :)

**Comment by pietrushnic** [comment]
When we improve Openess-Score, we should fix all graphics already published by Openess-Score v0.1. What is more, we should mark what version of Openess-Score was used.

**Comment by miczyg1** [comment]
Based on https://github.com/Dasharo/dasharo-issues/issues/1180#issuecomment-2890613780

We could show on two diagrams (assuming we now pass two binaries and dasharo templates will have to account for that):

- how much % or bytes has been liberated (i.e. closed-source proprietary minus closed-source Dasharo)
- how much % or bytes has been freed (i.e. empty space proprietary minus empty space Dasharo)

Another thing to consider is vboot. Accounting for redundant copies of the same blobs and open-source code may obfuscate the metrics in a good and bad way at the same time. So I think it may be wise to add `ignore-vboot` switch, which would skip analyzing RW_SECTION_A/RW_SECTION_B regions and assume they contain the same copies of components as in COREBOOT region, essentially treating RW_SECTION_A/RW_SECTION_B as empty space (as otherwise it would be without vboot). That switch should only be respected for comparison mode, rather for regular analysis fo the binary.

@pietrushnic let me know WDYT

**Comment by pietrushnic** [comment]
@miczyg1 agree. Concept-wise, it makes sense, the only thing would be **really nice graphics** we need here, something gleaming which we could use in marketing materials easily in every release. 

Does it make a difference for which vendor architecture the firmware was produced for?

**Comment by pietrushnic** [comment]
Yes, this makes sense. 

> UEFI images are parsed based on the UEFITool output of components in the image. Theoretically if someting can be parsed by UEFITool, the Openness Score will be able to deal with it, regardless of CPU architecture.

At some point, we should contribute to UEFITool's parsing of coreboot images with UEFI payload; then we could eliminate cbfstool. @NikolajSchlej do you see value in such support?

**Comment by miczyg1** [comment]
> Yes, this makes sense.
> 
> > UEFI images are parsed based on the UEFITool output of components in the image. Theoretically if someting can be parsed by UEFITool, the Openness Score will be able to deal with it, regardless of CPU architecture.
> 
> At some point, we should contribute to UEFITool's parsing of coreboot images with UEFI payload; then we could eliminate cbfstool. [@NikolajSchlej](https://github.com/NikolajSchlej) do you see value in such support?

coreboot images do not comply to UEFI firmware volume structures. I doubt there is any value to have non-UEFI image aprsing capability in a tool for UEFI images...

You can't eliminate cbfstool, it has many more purposes for coreboot build system.

**Comment by NikolajSchlej** [comment]
We sort of already lost the "we only parse UEFI PI-conformant images in UEFITool" as a goal, because all CPU vendors eventually ended up creating their own wrappers around UEFI PI structures, added flash descriptors, PSP directories and so on. Adding CBFS parser to the mix will not be an issue, but I also agree with @miczyg1 here about `cbfstool` deprecation - no need to do it, as UEFITool is a poor replacement for tools used by the build system.

**Comment by miczyg1** [comment]
> Ok, it makes sense at a different point and for reasons other than those mentioned here. I am already showing students the structure of Dasharo (coreboot+UEFI) binaries and getting the question why in the UEFI Tool it is not parsed, so it is good to know that potentially such a parser would be accepted.

UEFITool has a hard job parsing it because:

1. No UEFI Firmware Volume Top file present in coreboot images.
2. The whole UEFI Paylaod is LZMA compressed. Usually FVs which are LZMA compressed have a proper FFSv3 header. coreboto obviously doesn't do it. You may open `coreboot/build/UEFIPAYLOAD.fd` file in UEFITool easily though.

@pietrushnic

**Comment by miczyg1** [comment]
> Code that can generate a comparison table of various firmware already scored on the Dasharo page; I could present that on the next DUG because it could be interesting for the community.

Done in https://github.com/Dasharo/Openness-Score/pull/13

**Comment by DaniilKl** [comment]
@pietrushnic, @miczyg1, just to make sure I understood correctly. Should only the second idea be imeplemented?

The [first idea](https://github.com/Dasharo/dasharo-issues/issues/924#issuecomment-2548018166):

> Maybe we could simply introduce "liberation factor" calculated with the formula:
> 
> (open source bytes + empty bytes) x 100% / closed source bytes. That way we promote empty space in favor of open source code. We can add an exception: if open source bytes == 0 then liberation factor = 0%. I don't think overcomplicating it will benefit us at all.

The [second idea](https://github.com/Dasharo/dasharo-issues/issues/924#issuecomment-2890657749):

> We could show on two diagrams (assuming we now pass two binaries and dasharo templates will have to account for that):
> 
>     * how much % or bytes has been liberated (i.e. closed-source proprietary minus closed-source Dasharo)
> 
>     * how much % or bytes has been emptied/saved, TCB reduced (i.e. empty space Dasharo minus empty space proprietary)

**Comment by pietrushnic** [comment]
@DaniilKl yes, but demo/PoC over specification, unless we will see some mockups/demos/examples how that would work we cannot say which architecture of information will express our intent better.

**Comment by DaniilKl** [comment]
Ok, thank you for the confirmation.

**Comment by aronowski** [comment]
@DaniilKl @pietrushnic @miczyg1 Visual proof-of-concepts added below. Currently we have this visualization:

||
|-|
| <img width="480" height="360" alt="Image" src="https://github.com/user-attachments/assets/8f4d454e-b0ef-4409-9c5a-954d7de8c95e" /> |
| <img width="480" height="360" alt="Image" src="https://github.com/user-attachments/assets/071e3323-c905-4305-9e08-abf5b9833c8f" /> |

> The [first idea](https://github.com/Dasharo/dasharo-issues/issues/924#issuecomment-2548018166):
> 
> > Maybe we could simply introduce "liberation factor" calculated with the formula:
> > (open source bytes + empty bytes) x 100% / closed source bytes. That way we promote empty space in favor of open source code. We can add an exception: if open source bytes == 0 then liberation factor = 0%. I don't think overcomplicating it will benefit us at all.

| Liberation factor |
|-|
| <img width="480" height="360" alt="Image" src="https://github.com/user-attachments/assets/8f4d454e-b0ef-4409-9c5a-954d7de8c95e" /> |
| <img width="480" height="360" alt="Image" src="https://github.com/user-attachments/assets/09f58a8d-17ee-4663-b7c2-d5be3bf58828" /> |


> The [second idea](https://github.com/Dasharo/dasharo-issues/issues/924#issuecomment-2890657749):
> 
> > We could show on two diagrams (assuming we now pass two binaries and dasharo templates will have to account for that):
> > ```
> > * how much % or bytes has been liberated (i.e. closed-source proprietary minus closed-source Dasharo)
> > 
> > * how much % or bytes has been emptied/saved, TCB reduced (i.e. empty space Dasharo minus empty space proprietary)
> > ```

| Two-diagram proprietary vs Dasharo |
|-|
| <img width="480" height="360" alt="Image" src="https://github.com/user-attachments/assets/8f4d454e-b0ef-4409-9c5a-954d7de8c95e" /> |
| <img width="480" height="360" alt="Image" src="https://github.com/user-attachments/assets/86646c36-1f8f-4081-8472-6be42633af76" /> |
| <img width="480" height="360" alt="Image" src="https://github.com/user-attachments/assets/67221b7c-513a-4abe-a250-093c8ac71e2e" /> |
